duplex_http_call.go

```diff
@@ -15,55 +15,49 @@
 package connect
 
 import (
+	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"io"
 	"net/http"
 	"net/url"
 	"sync"
+	"sync/atomic"
 )
 
-// duplexHTTPCall is a full-duplex stream between the client and server. The
-// request body is the stream from client to server, and the response body is
-// the reverse.
-//
-// Be warned: we need to use some lesser-known APIs to do this with net/http.
-type duplexHTTPCall struct {
+// httpCall builds a HTTP request and sends it to a servcer. Unary calls block
+// until the response is received. Streaming calls return immediately.
+type httpCall struct {
 	ctx              context.Context
-	httpClient       HTTPClient
+	client           HTTPClient
 	streamType       StreamType
 	onRequestSend    func(*http.Request)
 	validateResponse func(*http.Response) *Error
 
-	// We'll use a pipe as the request body. We hand the read side of the pipe to
-	// net/http, and we write to the write side (naturally). The two ends are
-	// safe to use concurrently.
-	requestBodyReader *io.PipeReader
+	// io.Pipe is used to implement the request body for client streaming calls.
+	// If the request is unary, this is nil.
 	requestBodyWriter *io.PipeWriter
 
-	sendRequestOnce sync.Once
-	responseReady   chan struct{}
-	request         *http.Request
-	response        *http.Response
+	requestSent atomic.Bool
```

Comment:
Switch to `atomic.Bool` to be able to get the state of the `sync.Once` and error on unary if already called.